src/sidebar/services/api.js

```diff
@@ -175,35 +115,58 @@ function createAPICall(
           headers.Authorization = 'Bearer ' + token;
         }
 
-        const req = {
-          data: data ? stripInternalProperties(data) : null,
-          headers: headers,
+        const clientId = getClientId();
+        if (clientId) {
+          headers['X-Client-Id'] = clientId;
+        }
+
+        const { url, params: queryParams } = replaceURLParams(
+          descriptor.url,
+          params
+        );
+        const apiUrl = new URL(url);
+        apiUrl.search = queryString.stringify(queryParams);
+
+        return fetch(apiUrl.toString(), {
+          body: data ? JSON.stringify(stripInternalProperties(data)) : null,
+          headers,
           method: descriptor.method,
-          params: url.params,
-          paramSerializer: serializeParams,
-          url: url.url,
-        };
-        return $http(req);
+        });
       })
-      .then(function(response) {
-        onRequestFinished();
+      .then(response => {
+        let data;
 
-        if (options.includeMetadata) {
-          return { data: response.data, token: accessToken };
+        const status = response.status;
+        if (status >= 200 && status !== 204 && status < 500) {
+          data = response.json();
         } else {
-          return response.data;
+          data = response.text();
         }
+        return Promise.all([response, data]);
       })
-      .catch(function(response) {
-        onRequestFinished();
-
-        // Translate the API result into an `Error` to follow the convention that
-        // Promises should be rejected with an Error or Error-like object.
-        //
-        // Use `$q.reject` rather than just rethrowing the Error here due to
-        // mishandling of errors thrown inside `catch` handlers in Angular < 1.6
-        return $q.reject(translateResponseToError(response));
-      });
+      .then(
+        ([response, data]) => {
+          // `fetch` executed the request and the response was successfully parsed.
+          onRequestFinished();
+
+          if (response.status >= 400) {
+            // Translate the API result into an `Error` to follow the convention that
+            // Promises should be rejected with an Error or Error-like object.
+            throw translateResponseToError(response, data);
+          }
+
+          if (options.includeMetadata) {
```

Comment:
This `options.includeMetaData` thingy never fails to confuse me lightly :)

Reply:
Yeah. It turns out we still need it for the moment but we might be able to find a way to refactor it away in future.